src/libmongoc/tests/unified/entity-map.c

```diff
@@ -0,0 +1,603 @@
+   if (!client_entity) {
+      goto done;
+   }
+
+   client = (mongoc_client_t *) client_entity->value;
+   entity->value = (void *) mongoc_client_get_database (client, database_name);
+
+   ret = true;
+done:
+   bson_free (client_id);
+   bson_free (database_name);
+   bson_parser_destroy (parser);
+   if (!ret) {
+      entity_destroy (entity);
+      return NULL;
+   }
+   return entity;
+}
+
+entity_t *
+entity_collection_new (entity_map_t *entity_map,
+                       bson_t *bson,
+                       bson_error_t *error)
+{
+   bson_parser_t *parser;
+   entity_t *entity;
+   entity_t *database_entity;
+   mongoc_database_t *database;
+   bool ret = false;
+   char *database_id = NULL;
+   char *collection_name = NULL;
+
+   entity = entity_new ("collection");
+   parser = bson_parser_new ();
+   bson_parser_utf8 (parser, "id", &entity->id);
+   bson_parser_utf8 (parser, "database", &database_id);
+   bson_parser_utf8 (parser, "collectionName", &collection_name);
+   if (!bson_parser_parse (parser, bson, error)) {
+      goto done;
+   }
+
+   database_entity = entity_map_get (entity_map, database_id, error);
+   database = (mongoc_database_t *) database_entity->value;
+   entity->value = mongoc_database_get_collection (database, collection_name);
+   ret = true;
+done:
+   bson_free (collection_name);
+   bson_free (database_id);
+   bson_parser_destroy (parser);
+   if (!ret) {
+      entity_destroy (entity);
+      return NULL;
+   }
+   return entity;
+}
+
+mongoc_session_opt_t *
+session_opts_new (bson_t *bson, bson_error_t *error)
+{
+   bool ret = false;
+   mongoc_session_opt_t *opts = NULL;
+   bson_parser_t *parser;
+   bool *causal_consistency = NULL;
+
+   parser = bson_parser_new ();
+   bson_parser_bool_optional (parser, "causalConsistency", &causal_consistency);
+   if (!bson_parser_parse (parser, bson, error)) {
+      goto done;
+   }
+
+   opts = mongoc_session_opts_new ();
+   if (causal_consistency) {
+      mongoc_session_opts_set_causal_consistency (opts, *causal_consistency);
+   }
+
+done:
+   bson_parser_destroy (parser);
+   bson_free (causal_consistency);
+   if (!ret) {
+      mongoc_session_opts_destroy (opts);
+      return NULL;
+   }
+   return opts;
+}
+
+entity_t *
+entity_session_new (entity_map_t *entity_map, bson_t *bson, bson_error_t *error)
+{
+   bson_parser_t *parser;
+   entity_t *entity;
+   entity_t *client_entity;
+   mongoc_client_t *client;
+   char *client_id = NULL;
+   bson_t *session_opts_bson = NULL;
+   mongoc_session_opt_t *session_opts = NULL;
+   bool ret = false;
+
+   entity = entity_new ("session");
+   parser = bson_parser_new ();
+   bson_parser_utf8 (parser, "id", &entity->id);
+   bson_parser_utf8 (parser, "client", &client_id);
+   bson_parser_doc (parser, "sessionOptions", &session_opts_bson);
+   if (!bson_parser_parse (parser, bson, error)) {
+      goto done;
+   }
+
+   client_entity = entity_map_get (entity_map, client_id, error);
```

Comment:
In theory we could fail this lookup if a test makes a mistake with its entity list or defines entities out-of-order, right?  Let's just add a check here that would help us debug.

Reply:
Done.